node_exporter.go

```diff
@@ -27,51 +27,116 @@ import (
 	"gopkg.in/alecthomas/kingpin.v2"
 )
 
-func init() {
-	prometheus.MustRegister(version.NewCollector("node_exporter"))
+// handler wraps an unfiltered http.Handler but uses a filtered handler,
+// created on the fly, if filtering is requested. Create instances with
+// newHandler.
+type handler struct {
+	unfilteredHandler http.Handler
+	// exporterMetricsRegistry is a separate registry for the metrics about
+	// the exporter itself.
+	exporterMetricsRegistry *prometheus.Registry
+	includeExporterMetrics  bool
 }
 
-func handler(w http.ResponseWriter, r *http.Request) {
+func newHandler(includeExporterMetrics bool) *handler {
+	h := &handler{
+		exporterMetricsRegistry: prometheus.NewRegistry(),
+		includeExporterMetrics:  includeExporterMetrics,
+	}
+	if h.includeExporterMetrics {
+		h.exporterMetricsRegistry.MustRegister(
+			prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
```

Comment:
Removing this seems a bit much, we get reasonably regular bug reports about fd leaks and memory issues and this could leave us blind.

Reply:
But isn't that the point here? Those metrics are not removed by default, and we generally recommend to keep them to diagnose issues of the exporter itself, but those who explicitly want to disable the metrics to save scraping and storing them should be able to do so. If they see issues, they have to re-activate the metrics.

Reply:
Yes, this is the point.